src/java/org/apache/cassandra/journal/OnDiskIndex.java

```diff
@@ -0,0 +1,306 @@
+        if (keyIndex < 0)
+            return EMPTY;
+
+        int[] offsets = new int[] { offsetAtIndex(keyIndex) };
+
+        /*
+         * Duplicate entries are possible within one segment (but should be rare).
+         * Check and add entries before and after the found result (not guaranteed to be first).
+         */
+
+        for (int i = keyIndex - 1; i >= 0 && id.equals(keyAtIndex(i)); i--)
+        {
+            int length = offsets.length;
+            offsets = Arrays.copyOf(offsets, length + 1);
+            offsets[length] = offsetAtIndex(i);
+        }
+
+        for (int i = keyIndex + 1; i < entryCount && id.equals(keyAtIndex(i)); i++)
+        {
+            int length = offsets.length;
+            offsets = Arrays.copyOf(offsets, length + 1);
+            offsets[length] = offsetAtIndex(i);
+        }
+
+        Arrays.sort(offsets);
+        return offsets;
+    }
+
+    @Override
+    public int lookUpFirst(K id)
+    {
+        if (!mayContainId(id, keySupport))
+            return -1;
+
+        int keyIndex = binarySearch(id);
+
+        /*
+         * Duplicate entries are possible within one segment (but should be rare).
+         * Check and add entries before until we find the first occurrence of key.
+         */
+        for (int i = keyIndex - 1; i >= 0 && id.equals(keyAtIndex(i)); i--)
+            keyIndex = i;
+
+        return keyIndex < 0 ? -1 : offsetAtIndex(keyIndex);
+    }
+
+    private K keyAtIndex(int index)
+    {
+        return keySupport.deserialize(buffer, FILE_PREFIX_SIZE + index * ENTRY_SIZE, descriptor.userVersion);
+    }
+
+    private int offsetAtIndex(int index)
+    {
+        return buffer.getInt(FILE_PREFIX_SIZE + index * ENTRY_SIZE + KEY_SIZE);
+    }
+
+    /*
+     * This has been lifted from {@see IndexSummary}'s implementation,
+     * which itself was lifted from Harmony's Collections implementation.
+     */
+    private int binarySearch(K key)
+    {
+        int low = 0, mid = entryCount, high = mid - 1, result = -1;
+        while (low <= high)
+        {
+            mid = (low + high) >> 1;
+            result = -compareWithKeyAt(key, mid);
+            if (result > 0)
+            {
+                low = mid + 1;
+            }
+            else if (result == 0)
+            {
+                return mid;
+            }
+            else
+            {
+                high = mid - 1;
+            }
+        }
+        return -mid - (result < 0 ? 1 : 2);
+    }
+
+    private int compareWithKeyAt(K key, int keyIndex)
+    {
+        int offset = FILE_PREFIX_SIZE + ENTRY_SIZE * keyIndex;
+        return keySupport.compareWithKeyAt(key, buffer, offset, descriptor.userVersion);
+    }
+
+    @Override
+    public void update(K id, int offset)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void persist(Descriptor descriptor)
+    {
+        throw new UnsupportedOperationException();
+    }
```

Comment:
It would be, I guess, but it's done this way so `Segment` logic could be shared between `ActiveSegment` and `StaticSegment` without too much hassle. I think it's okayish on balance.

Reply:
Taking a quick look, if `Segment` was defined as

```
abstract class Segment<K, I extends Index<K>, Self extends Segment<K, I, Self>> implements Closeable, RefCounted<Self>
```

then this is all handled at the type system, and the code change is small.

Reply:
Take a look at https://github.com/iamaleksey/cassandra/pull/2/files, I did the work to add `MutableIndex`, the biggest impact to the diff is due to `Ref`

Reply:
Did a simpler version of that change in most recent commit, in a way that still addresses the raised concern, cheers.